### ISSUE TITLE
Large memory is required

### ISSUE THREAD

**Issue by evayfang2019** [mcgilldinglab/MATES#13]
```python
data_processor.generate_training_sample(data_mode, sample_list_file, bin_size, proportion)
# data_mode: Smart_seq 
# bin_size: 5 
# proportion: 80
```

When I analyzed smart-seq data (2377 cells), this step required a large memory (more than 750GB). 
Reporting:
`numpy.core._exceptions._ArrayMemoryError: Unable to allocate 1.71 GiB for an array with shape (114806, 2001) and data type float64`

How should I solve it?

**Comment by Szym29** [comment]
Hi @evayfang2019 ,

Thank you for using our tool. We will investigate this issue using the smart-seq data used in our manuscript to see if there is any potential issue. 

In the meanwhile, would you mind sharing more detailed Traceback of the error? It would be great if you can share the bam file of one or two cells of your data along with other input files like the type of species, barcode.tsv, sample_list_file.txt, etc. Then we can directly fix the issue based on your data.

Thanks,
Yumin

**Comment by Szym29** [comment]
Hi @evayfang2019 ,

Thank you for your reply! Please send the files to yumin.zheng@mail.mcgill.ca. Or you can upload it to cloud storage service like google drive, onedrive. 

Thanks,
Yumin

**Comment by Szym29** [comment]
Hi @evayfang2019 ,

We tested the sample data you shared with us and it looks fine in our end and their memory cost is reasonable. Have these two bam files been sub-sampled? Ideally, MATES will process Smart-seq's files one by one, would you mind trying again and share the bam file that causes the OOM issue? 
To find which bam file cause this issue, you can add `print(cb_path)` before the line 80 in the [./MATES/scripts/generateTraining.py](https://github.com/mcgilldinglab/MATES/blob/main/MATES/scripts/generateTraining.py#L80). Then run `pip install .` again to install the MATES. And run the `data_processor.generate_training_sample` again. Once the MATES stop running, you can find the bam file that cause this error. 

Below is our scripts to run the bam files you shared with us. You can check if there it is different from your script.

```python
from MATES import bam_processor, data_processor, MATES_model
bam_processor.split_bam_files('Smart_seq', 2, 'sample_list.txt', 'bam_files_dir.txt')
bam_processor.count_coverage_vec('exclusive', 'Smart_seq', 2, 'sample_list.txt', ref_path="TE_nooverlap.csv")
data_processor.calculate_UM_region('exclusive', 'Smart_seq', 'sample_list.txt', bin_size=5, proportion=10,cut_off=0)
data_processor.generate_training_sample('Smart_seq', 'sample_list.txt', bin_size=5, proportion=10,cut_off=1)
data_processor.generate_prediction_sample('exclusive','Smart_seq','sample_list.txt', bin_size=5, proportion=10, ref_path="TE_nooverlap.csv",cut_off=1)

##Step4:
MATES_model.train('Smart_seq', 'sample_list.txt', bin_size = 5, proportion = 10, BATCH_SIZE= 4096, 
                  AE_LR = 1e-4, MLP_LR = 1e-6, AE_EPOCHS = 10, MLP_EPOCHS = 10, DEVICE='cuda:0')
MATES_model.prediction('exclusive', 'Smart_seq', 'sample_list.txt', bin_size = 5, proportion = 10, 
                       AE_trained_epochs =10, MLP_trained_epochs=10,DEVICE='cuda:0')
```
[bam_files_dir.txt](https://github.com/user-attachments/files/17565459/bam_files_dir.txt)
[sample_list.txt](https://github.com/user-attachments/files/17565460/sample_list.txt)
I also attached the bam_files_dir and sample_list files here.

Thanks,
Yumin

**Comment by Szym29** [comment]
Hi @evayfang2019 ,

The proportion threshold is used to define the multi-mapping and uniquely mapping regions. In our manuscript, we have the following description. Please our [manuscript](https://www.nature.com/articles/s41467-024-53114-7) for more details. 

> Selection of the hyperparameters, the bin size (W) and proportion threshold (P), is dependent on the data’s multi-mapping rates. For data with high multi-mapping rates, we opt for a smaller bin size and a higher threshold to enhance accuracy. Conversely, for low multi-mapping rate data, we prefer a larger bin size and a lower threshold to acquire more training samples (please see Supplementary Figs. S19d, f for the selection of those parameters).   

But we do not encourage you to use too low threshold, which would limited the performance of model training. 

For the batch_size error, I think it had been fixed in the previous version. You can update the MATES to v0.1.3. But I will double check it to make sure it has been resolved. 

For the downstream analysis, the index can be refereed back to the TE reference bed file you send to MATES.  For example the 'TE_nooverlap.csv' in my previous script. Currently, you can use 
`pd.read_csv('TE_nooverlap.csv',header=None,names=['chromosome', 'start', 'end', 'TE_Name', 'index', 'strand','TE_fam', 'length'])` to open the TE reference file and the TE_index can be matched with the 'index' column. Then you can use 'chromosome|start|end|TE_Name' to rename it.

For the last point, we are working on it to provide a function that can directly generate the specific locus TE names as you wanted. This function can be finished in one or two weeks and I will let you know once we finished it. 

Thanks,
Yumin

**Comment by Szym29** [comment]
Hi @evayfang2019 ,

In MATES v0.1.5, we have supported to generate h5ad files with var name as the specific TE names with chromosome, start-end, TE name, TE family information.

Here are some example codes,

```python
from MATES import TE_quantifier, MATES_model
MATES_model.train('Smart_seq', 'sample_list.txt', bin_size = 5, proportion = 10, BATCH_SIZE= 512, AE_LR = 1e-4, MLP_LR = 1e-6, AE_EPOCHS = 10, MLP_EPOCHS = 10, DEVICE='cuda:0')
MATES_model.prediction_locus('exclusive', 'Smart_seq', 'sample_list.txt', bin_size = 5, proportion = 10,  AE_trained_epochs =10, MLP_trained_epochs=10,DEVICE='cuda:0')
TE_quantifier.quantify_locus_TE_MTX('exclusive', 'Smart_seq', 'sample_list.txt')
```

The outputed h5ad file will be saved to `./Smartseq_locus/combined_matrix.h5ad`. You can concatenate it to the gene expression data. 

Thanks,
Yumin

**Comment by XiaoyuZhan520** [comment]
Tasks set to Proportion=10 were still killed.
